src/aspire/abinitio/commonline_d2.py

```diff
@@ -0,0 +1,1888 @@
+        )
+        i_idx = np.repeat(np.arange(self.n_inplane_rots), self.n_inplane_rots // 2)
+        j_idx = np.tile(np.arange(self.n_inplane_rots // 2), self.n_inplane_rots)
+        idx_vec = np.arange(n_rot_1)
+        idx = 0
+
+        for i in range(n_rot_1):
+            unique_pairs_i = idx_vec[self.eq2eq_Rij_table_11[i]]
+            if len(unique_pairs_i) == 0:
+                continue
+            i_idx_plus_offset = i_idx + (i * self.n_inplane_rots)
+
+            for j in unique_pairs_i:
+                j_idx_plus_offset = j_idx + (j * self.n_inplane_rots)
+                oct1_ij_map[:, :, idx] = np.column_stack(
+                    (i_idx_plus_offset, j_idx_plus_offset)
+                )
+                idx += 1
+
+        # Now the map for i<j pairs for Ri in octant 1 and Rj in octant 2.
+        n_pairs_12 = np.sum(self.eq2eq_Rij_table_12)
+        oct2_ij_map = np.zeros(
+            (self.n_inplane_rots**2 // 2, 2, n_pairs_12), dtype=np.int64
+        )
+        idx_vec = np.arange(n_rot_2)
+        idx = 0
+
+        for i in range(n_rot_1):
+            unique_pairs_i = idx_vec[self.eq2eq_Rij_table_12[i] > 0]
+            if len(unique_pairs_i) == 0:
+                continue
+            i_idx_plus_offset = i_idx + (i * self.n_inplane_rots)
+
+            for j in unique_pairs_i:
+                j_idx_plus_offset = j_idx + (j * self.n_inplane_rots)
+                oct2_ij_map[:, :, idx] = np.column_stack(
+                    (i_idx_plus_offset, j_idx_plus_offset)
+                )
+                idx += 1
+
+        tmp1 = oct1_ij_map[:, 0, :]
+        tmp2 = oct1_ij_map[:, 1, :]
+        self.oct1_ij_map = np.column_stack(
+            (tmp1.flatten(order="F"), tmp2.flatten(order="F"))
+        )
+
+        tmp1 = oct2_ij_map[:, 0, :]
+        tmp2 = oct2_ij_map[:, 1, :]
+        self.oct2_ij_map = np.column_stack(
+            (tmp1.flatten(order="F"), tmp2.flatten(order="F"))
+        )
+
+    ##############################################
+    # Compute Self-Commonline Correlation Scores #
+    ##############################################
+
+    def _compute_scl_scores(self):
+        """
+        Compute correlations for self-commonline candidates.
+        """
+        logger.info("Computing self-commonline correlation scores.")
+        n_img = self.n_img
+        n_theta = self.n_theta
+        n_eq = len(self.non_tv_eq_idx)
+        n_inplane = self.n_inplane_rots
+
+        # Run ML in parallel
+        scl_matrix = np.concatenate((self.scl_idx_1, self.scl_idx_2))
+        M = len(scl_matrix) // 3
+        corrs_out = np.zeros((n_img, M), dtype=self.dtype)
+        scl_idx = scl_matrix.reshape(M, 3)
+
+        # Get non-equator indices to use with corrs matrix.
+        non_eq_lin_idx = self.non_eq_idx.flatten()
+        n_non_eq = len(non_eq_lin_idx)
+        non_eq_idx = np.unravel_index(
+            scl_idx[non_eq_lin_idx].flatten(), (n_theta // 2, n_theta)
```

Comment:
No, `scl_idx` is shape (M, 3) and we're indexing into the first axis with `non_eq_lin_idx`, so `scl_idx[non_eq_lin_idx]` will be shape (`len(non_eq_lin_idx)`, 3).

Reply:
Cool. My mistake; I thought this was one of those cases where numpy auto-flattens when given arbitrary indices, but  happens only for more than 1 d (because could be ragged). Here you just have the one. Got it. Thanks.